plugins/synaprom/fu-synaprom-device.c

```diff
@@ -0,0 +1,504 @@
+		if (rep_hdr->status != 0) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "failed to send command: error status: 0x%x",
+				     rep_hdr->status);
+			return FALSE;
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static GByteArray *
+fu_synaprom_request_new (guint8 cmd, gsize cmdlen)
+{
+	GByteArray *blob = g_byte_array_new ();
+	g_byte_array_set_size (blob, cmdlen);
+	blob->data[0] = cmd;
+	return blob;
+}
+
+static GByteArray *
+fu_synaprom_reply_new (gsize cmdlen)
+{
+	GByteArray *blob = g_byte_array_new ();
+	g_byte_array_set_size (blob, cmdlen);
+	return blob;
+}
+
+static gboolean
+fu_synaprom_device_setup (FuDevice *device, GError **error)
+{
+	FuSynapromDevice *self = FU_SYNAPROM_DEVICE (device);
+	FuSynapromReplyGetVersion pkt;
+	g_autoptr(GByteArray) request = NULL;
+	g_autoptr(GByteArray) reply = NULL;
+	g_autofree gchar *version = NULL;
+
+	/* get version */
+	request = fu_synaprom_request_new (FU_SYNAPROM_CMD_GET_VERSION, sizeof(FuSynapromCmdGeneric));
+	reply = fu_synaprom_reply_new (sizeof(FuSynapromReplyGetVersion));
+	if (!fu_synaprom_device_cmd_send (self, request, reply, 250, error)) {
+		g_prefix_error (error, "failed to get version: ");
+		return FALSE;
+	}
+	memcpy (&pkt, reply->data, sizeof(pkt));
+	g_debug ("product ID is %u, version=%u.%u, buildnum=%u prod=%i",
+		 pkt.product, pkt.vmajor, pkt.vminor, pkt.buildnum,
+		 pkt.security[1] & FU_SYNAPROM_SECURITY1_PROD_SENSOR);
+	version = g_strdup_printf ("%02u.%02u.%u",
+				   pkt.vmajor, pkt.vminor, pkt.buildnum);
+	fu_device_set_version (device, version, FWUPD_VERSION_FORMAT_TRIPLET);
+
+	/* we need this for checking the firmware later */
+	self->vmajor = pkt.vmajor;
+	self->vminor = pkt.vminor;
+
+	/* check device type */
+	if (pkt.product == FU_SYNAPROM_PRODUCT_PROMETHEUS) {
+		fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	} else if (pkt.product == FU_SYNAPROM_PRODUCT_PROMETHEUSPBL ||
+		   pkt.product == FU_SYNAPROM_PRODUCT_PROMETHEUSMSBL) {
+		fu_device_add_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	} else {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "device %u is not supported by this plugin",
+			     pkt.product);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaprom_device_cmd_download_chunk (FuSynapromDevice *device,
+				       const GByteArray *chunk,
+				       GError **error)
+{
+	g_autoptr(GByteArray) request = NULL;
+	g_autoptr(GByteArray) reply = NULL;
+	request = fu_synaprom_request_new (FU_SYNAPROM_CMD_BOOTLDR_PATCH,
+					   sizeof(FuSynapromCmdGeneric) + chunk->len);
+	memcpy (request->data + sizeof(FuSynapromCmdGeneric), chunk->data, chunk->len);
+	reply = fu_synaprom_reply_new (sizeof(FuSynapromReplyGeneric));
+	return fu_synaprom_device_cmd_send (device, request, reply, 20000, error);
+}
+
+static GBytes *
+fu_synaprom_device_prepare_firmware (FuDevice *device,
```

Comment:
This function does a lot of good sanity checks on the firmware.  What would you think about adding a call to it to the test suite?  It can help to verify that a firmware payload will pass fwupd's checks before even putting it into fwupd then.

Reply:
Good idea, will do. I also need to add CONFIG type firmware, which means restructuring this a bit anyway.